include/tscore/ver_ptr.h

```diff
@@ -0,0 +1,288 @@
+      using Atomic_access_type = AAT;
+
+      explicit Type_(void *p = nullptr, unsigned v = 0)
+      {
+        _vp.ptr     = p;
+        _vp.version = v;
+
+        const std::size_t Pad = sizeof(_aa) - sizeof(_vp);
+        if (Pad) {
+          // Make sure pad bytes are consistently zero.
+          //
+          std::memset(reinterpret_cast<char *>(&_vp) + sizeof(_vp), 0, Pad);
+        }
+      }
+
+      explicit Type_(Atomic_access_type const &aa) { _aa = aa; }
+
+      void *
+      ptr() const
+      {
+        return _vp.ptr;
+      }
+
+      void
+      ptr(void *p)
+      {
+        _vp.ptr = p;
+      }
+
+      unsigned
+      version() const
+      {
+        return _vp.version;
+      }
+
+      void
+      version(unsigned v)
+      {
+        _vp.version = v;
+      }
+
+    private:
+      union {
+        VPS _vp;
+        AAT _aa;
+      };
+
+      static_assert(sizeof(_vp) <= sizeof(_aa));
+    };
+
+    namespace
+    {
+      unsigned const Version_bits  = 15;
+      unsigned const Version_shift = 48;
+      constexpr unsigned
+      Mask_version(unsigned v)
+      {
+        return v & ((1U << (Version_bits + 1)) - 1);
+      }
+      std::uint64_t const Version_mask = static_cast<std::uint64_t>(Mask_version(~0U)) << Version_shift;
+    } // namespace
+
+    template <> class Type_<std::uint64_t, true>
+    {
+    public:
+      using Atomic_access_type = std::uint64_t;
+
+      Type_(){};
+
+      explicit Type_(void *p, unsigned v = 0)
+      {
+        _ptr = reinterpret_cast<std::uint64_t>(p) & ~Version_mask;
+        _ptr |= static_cast<uint64_t>(Mask_version(v)) << Version_shift;
+      }
+
+      explicit Type_(Atomic_access_type const &ptr) { _ptr = ptr; }
+
+      void *
+      ptr() const
+      {
+        return reinterpret_cast<void *>(_ptr & ~Version_mask);
+      }
+
+      void
+      ptr(void *p)
+      {
+        _ptr &= Version_mask;
+        _ptr |= reinterpret_cast<std::uint64_t>(p) & ~Version_mask;
+      }
+
+      unsigned
+      version() const
+      {
+        return Mask_version(static_cast<unsigned>(_ptr >> Version_shift));
+      }
+
+      void
+      version(unsigned v)
+      {
+        _ptr &= ~Version_mask;
+        _ptr |= static_cast<std::uint64_t>(Mask_version(v)) << Version_shift;
+      }
+
+    private:
+      std::uint64_t _ptr;
+    };
+
+    enum Impl_type {
+      DEFAULT,        // Atomic access may not be lock free.
+      VERSION_IN_PTR, // Use bits inside of 64-bit pointer to store version.
+      BITS64,         // VPS structure is less than 64 bits, and 64-bit access is lock free.
+      BITS128         // VPS structure is less than 128 bits, and 128-bit compare-and-exchange is available.
+    };
+
+    constexpr Impl_type
+    Impl()
+    {
+      if (std::atomic<std::uint64_t>::is_always_lock_free) {
+        if (sizeof(VPS) <= 8) {
+          return BITS64;
+
+        } else if (Ptr64bits15unused && (sizeof(void *) == 8)) {
+          return VERSION_IN_PTR;
```

Comment:
For 64-bit architectures, I am favoring the approach where the version in stored in ignored bits in the pointer.  Currently, ATS favors using 128-bit compare-and-swap/exchange if TS_HAS_128BIT_CAS is set to 1 in ink_config.h ( https://github.com/apache/trafficserver/blob/472075bd7b5755cc1d8bf9a4bbf750474c4da53c/include/tscore/ink_queue.h#L136 ).  But I think this flag may be set to 1 even if the 128-bit CAS is not atomic.  Looking at  this example code:  https://godbolt.org/z/ncYsv9 , the 128-bit CAS results in a call to a runtime library function (which I suspect contains a lock).

Interestingly, the 128-bit CAS translates directly to a HW instruction for ARM v8:  https://godbolt.org/z/aennnM .

Reply:
Hmm when I build master branch TS on my x86-64 MacBook, it does appear to somehow generate HW 128 bit CAS operations ( lock cmpxchg16b ) when I look at the objdump output.  So, I looks like I doubted TS in vain, I will change to favoring 128 bit CAS.

Reply:
Using the stolen pointer bits approach, with my new implementation, a regression test fails:  
```
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
RPRINT HostDBTests: HostDBTestReverse: reversed static-css-csq-225018.business.bouyguestelecom.com
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
RPRINT HostDBTests: HostDBTestReverse: reversed apn-77-114-28-193.dynamic.gprs.plus.pl
RPRINT HostDBTests: HostDBTestReverse: reversed 59-113-175-47.dynamic-ip.hinet.net
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
RPRINT HostDBTests: HostDBTestReverse: reversed 110-132-7-105.rev.home.ne.jp
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
RPRINT HostDBTests: HostDBTestReverse: reversed 165.sub-97-204-116.myvzw.com
RPRINT HostDBTests: HostDBTestReverse: reversed cpe-66-66-7-152.rochester.res.rr.com
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
RPRINT HostDBTests: HostDBTestReverse: reversed host88.fibrewired.on.ca
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
RPRINT HostDBProcessor: hostdbinfo r=a49f9780
RPRINT HostDBProcessor: hostdbinfo hostname=www.ibm.com
RPRINT HostDBProcessor: hostdbinfo rr 0
RPRINT HostDBProcessor: hostdbinfo A ip=23.47.163.207
RPRINT HostDBTests: HostDBTestReverse: reversed ec2-35-166-27-40.us-west-2.compute.amazonaws.com
RPRINT HostDBProcessor: hostdbinfo r=a49f9700
RPRINT HostDBProcessor: hostdbinfo hostname=www.microsoft.com
RPRINT HostDBProcessor: hostdbinfo rr 0
RPRINT HostDBProcessor: hostdbinfo A ip=23.37.117.182
Regression test(HostDBProcessor) still in progress
Regression test(HostDBTests) still in progress
Regression test(PVC) still in progress
RPRINT HostDBProcessor: hostdbinfo r=a49f9680
RPRINT HostDBProcessor: hostdbinfo hostname=www.coke.com
RPRINT HostDBProcessor: hostdbinfo rr 0
RPRINT HostDBProcessor: hostdbinfo A ip=104.121.207.153
RPRINT HostDBProcessor: hostdbinfo r=a49f9600
RPRINT HostDBProcessor: hostdbinfo hostname=(null)
RPRINT HostDBProcessor: hostdbinfo rr 0
RPRINT HostDBProcessor: hostdbinfo A ip=4.2.2.2
RPRINT HostDBProcessor: hostdbinfo r=a49f9580
RPRINT HostDBProcessor: hostdbinfo hostname=(null)
RPRINT HostDBProcessor: hostdbinfo rr 0
RPRINT HostDBProcessor: hostdbinfo A ip=127.0.0.1
RPRINT HostDBProcessor: HostDBTestRR: 0 outstanding 5 success 1 failure
RPRINT HostDBTests: HostDBTestReverse: reversed 108-198-194-205.lightspeed.jcsnms.sbcglobal.net
RPRINT HostDBTests: HostDBTestReverse: reversed host137-66-252-76.netaspx.com
    REGRESSION_RESULT HostDBProcessor:                          FAILED
```